Doc/importNpmLock: general improvements (#340019)
NixOS/nixpkgs: doc/languages-frameworks/javascript.section.md

#### importNpmLock {#javascript-buildNpmPackage-importNpmLock}

The `importNpmLock` function is a simpler and more convenient alternative to `fetchNpmDeps` for managing NPM dependencies in Nix. Unlike `fetchNpmDeps`, there is no need to specify a `hash` when using `importNpmLock`. This function relies entirely on the integrity hashes already present in the `package-lock.json` file.

> Here's what `importNpmLock` does:
> 
> - It processes the `package.json` and `package-lock.json` files from your NPM project.
> - It replaces the dependency references in these files with paths to the Nix store.
> - It allows you to customize how each dependency is fetched. See argument: `fetcherOpts`.

Reviewer: This should be the first sentence paragraph.

```suggestion
Replace the npm dependency references in `package.json` and `package-lock.json` with paths to the Nix store.
How each dependency is fetched can be customized with the `fetcherOpts` argument.
```

Although "The X function..." is used everywhere in this section, I really don't think it's a good pattern. It's pretty much an empty phrase because the context is immediately clear from the heading.

`importNpmLock` is a Nix function that requires the following optional arguments:

- `npmRoot`: Path to package directory containing the source tree. If this is omited `package` and `packageLock` arguments must be specified instead.
- `package`: Parsed contents of `package.json`
- `packageLock`: Parsed contents of `package-lock.json`
- `pname`: Package name
- `version`: Package version

> - `fetcherOpts`: An attribute set of arguments forwarded to the underlying fetcher.

Reviewer: ```suggestion
- `fetcherOpts`: An attribute set of arguments forwarded to the underlying [fetcher](#chap-pkgs-fetchers).
```

#### importNpmLock {#javascript-buildNpmPackage-importNpmLock}

This function replaces the npm dependency references in `package.json` and `package-lock.json` with paths to the Nix store.
How each dependency is fetched can be customized with the `fetcherOpts` argument.

> This is simpler and more convenient alternative to [`fetchNpmDeps`](#javascript-buildNpmPackage-fetchNpmDeps) for managing npm dependencies in Nix.

Reviewer: ```suggestion
This is a simpler and more convenient alternative to [`fetchNpmDeps`](#javascript-buildNpmPackage-fetchNpmDeps) for managing npm dependencies in Nixpkgs.
```

> There is no need to specify a `hash`; since it relies entirely on the integrity hashes already present in the `package-lock.json` file.

Reviewer: ```suggestion
There is no need to specify a `hash`, since it relies entirely on the integrity hashes already present in the `package-lock.json` file.
```

  npmConfigHook = importNpmLock.npmConfigHook;
}
```
:::

:::{.example}
##### `pkgs.importNpmLock` usage example with `fetcherOpts` {#javascript-buildNpmPackage-example-fetcherOpts}

`importNpmLock` uses the following fetchers:

- `pkgs.fetchurl` for `http(s)` dependencies
- `builtins.fetchGit` for `git` dependencies

It is possible to provide additional arguments as needed:

```nix

>   # ... omited for clarity

Reviewer: No, please make a self-contained example. We must expect those will be viewed completely out of context, and most often people don't even read the prose but just copy things around.

Reviewer: Sphinx has [line highlighting](https://mystmd.org/guide/code#numbering-and-highlighting), maybe we should add something like this to our system eventually.
